ouroboros-consensus/src/Ouroboros/Consensus/Util/CBOR.hs

```diff
@@ -158,42 +162,57 @@ readIncrementalOffsets
      --     (the size of @a@, and @a@ itself)),
      --     error encountered during deserialisation and the offset after
      --     which it occurred)
-readIncrementalOffsets hasFS@HasFS{..} decoder fp = withLiftST $ \liftST ->
-    withFile hasFS fp ReadMode $ \h -> do
-      fileSize <- hGetSize h
-      if fileSize == 0
-        -- If the file is empty, we will immediately get "end of input"
-        then return ([], Nothing)
-        else liftST (CBOR.deserialiseIncremental decoder) >>=
-             go liftST h 0 [] Nothing [] fileSize
+readIncrementalOffsets hasFS decoder fp =
+    withStreamIncrementalOffsets hasFS decoder fp $ \stream ->
+      (\(as :> mbErr) -> (as, mbErr)) <$> S.toList stream
+
+-- | Streaming variant of 'readIncrementalOffsets'
+--
+-- Continuation-passing style to ensure proper closure of the file.
+withStreamIncrementalOffsets
+  :: forall m h a r. (IOLike m, HasCallStack)
+  => HasFS m h
+  -> (forall s . CBOR.Decoder s (LBS.ByteString -> a))
+  -> FsPath
+  -> (Stream (Of (Word64, (Word64, a))) m (Maybe (ReadIncrementalErr, Word64)) -> m r)
```

Comment:
Minor sad face about the use of `streaming` (the package) rather than `conduit`. Let's add a comment here that if we introduce user-facing streaming API also, the fact that we are using `streaming` here should not dictate that we should stick with it later; rather, we should revisit this code at that point.